### PR TITLE
Implement vjps of vjps for HPE and SAP (CPU)

### PR DESCRIPTION
**Pull request by frostedoyster**
This is needed to train models on forces and stresses, for example

### CODE REVIEW

File: mops/src/hpe/cpu.tpp [hunk 1]
```diff
@@ -277,3 +270,247 @@ void mops::homogeneous_polynomial_evaluation_vjp(
 
     throw std::runtime_error("Only up to polynomial order 10 is supported at the moment. Please contact the developers for more");
 }
+
+
+template<typename scalar_t, uint8_t polynomial_order>
+void _homogeneous_polynomial_evaluation_vjp_vjp_templated_polynomial_order(
```
Comment (Luthaf) [inline mops/src/hpe/cpu.tpp:276]: this function is pretty big, and without any explanation. Could you add comments about what's happening (not line by line, but explaining the overall steps of the code?)

File: mops/src/sap/cpu.tpp [hunk 2]
```diff
@@ -228,38 +216,63 @@ void mops::sparse_accumulation_of_products_vjp(
...
-        grad_output_i = remainder_grad_o_ptr;
-        grad_a_i = remainder_grad_a_ptr;
-        grad_b_i = remainder_grad_b_ptr;
-        a_i = remainder_a_ptr;
-        b_i = remainder_b_ptr;
-        for (size_t i = 0; i < size_remainder; i++){
-            for (size_t j = 0; j < c_size; j++) {
-                scalar_t grad_output_j = grad_output_i[p_o_ptr[j]];
-                scalar_t common_factor = grad_output_j * c_ptr[j];
+        // // Handle the remainder, i.e., the elements that do not fit inside a multiple of simd_element_count
```
Comment (Luthaf) [inline mops/src/sap/cpu.tpp:221]: if this code is no longer needed, please remove it

File: mops/src/sap/cpu.tpp [hunk 3]
```diff
@@ -271,3 +284,269 @@ void mops::sparse_accumulation_of_products_vjp(
...
+    Tensor<scalar_t, 2> grad_grad_B,
+    Tensor<scalar_t, 2> grad_output,
+    Tensor<scalar_t, 2> A,
+    Tensor<scalar_t, 2> B,
+    Tensor<scalar_t, 1> C,
+    Tensor<int32_t, 1> indices_A,
+    Tensor<int32_t, 1> indices_B,
+    Tensor<int32_t, 1> indices_output
+) {
+    check_sap_vjp_vjp(grad_grad_output, grad_A_2, grad_B_2, grad_grad_A, grad_grad_B, grad_output, A, B, C, indices_A, indices_B, indices_output, "cpu_sparse_accumulation_of_products_vjp_vjp");
```
Comment (Luthaf) [inline mops/src/sap/cpu.tpp:303]: can we get some more explanations of what's going on inside here as well?

File: python/mops-torch/tests/sasaw.py [hunk 4]
```diff
@@ -51,35 +57,59 @@ def test_sasaw():
...
-    B = torch.rand(70, 50, dtype=torch.float64, requires_grad=True)
-    W = torch.rand(3, 5, 50, dtype=torch.float64, requires_grad=True)
-    C = torch.rand(50, dtype=torch.float64)
-    indices_output_1 = torch.randint(3, size=(70,), dtype=torch.int32)
-    indices_W_1 = torch.randint(3, size=(70,), dtype=torch.int32)
-    output_size_2 = 15
-    indices_A = torch.randint(10, size=(50,), dtype=torch.int32)
-    indices_W_2 = torch.randint(5, size=(50,), dtype=torch.int32)
-    indices_output_2 = torch.randint(output_size_2, size=(50,), dtype=torch.int32)
+@pytest.mark.parametrize("dtype", [torch.float64])
```
Comment (Luthaf) [inline python/mops-torch/tests/sasaw.py:60]: We don't need to parametrize if there is a single value, just use this dtype in the test code

**Review by Luthaf — COMMENTED** [review]
This looks good overall, but I don't understand what's going on for the actual implementations

**Comment by Luthaf** [comment]
Regarding the test failure: this could be caused by a data race in OpenMP. You could try running the tests without OpenMP to check if that's the case, and/or compiling the code with the [threads sanitizer](https://github.com/google/sanitizers/wiki/ThreadSanitizerCppManual), which should catch some issues at runtime.

**Comment by frostedoyster** [comment]
An update on the bug:

Thanks @Luthaf for the suggestion to look into parallelism and use thread sanitizer. That was indeed the issue because removing the OpenMP parallelism from HPE indeed eliminates the test failure. However, I found it to be nearly impossible to test the code with thread sanitizer: on my laptop (Ubuntu 23.10), there is a known bug that prevents thread sanitizer to work properly ("unknown memory mapping"), on CI (ubuntu 20.04) there is another known bug for which the compiled thread sanitizer library is missing (linker fails), on Jed mops is very difficult to build in the first place because it requires a workaround (due to the conflict of Jed having a cuda compiler and `libtorch_cuda.so` missing).

To that, I should add that the double backward of HPE is virtually useless for us: HPE cannot be used during training of equivariant models (where SAP would take its place). However, if one uses ACE/WK, the trained model can be re-expressed in terms of a few call to HPE (rather than many to SAP), and that is its main use case. As a result, the double backward for HPE is there just for completeness, but it is virtually useless in our field (and I suspect in general... who's going to need VJP of VJP of homogeneous polynomials?).

As a result, I disabled parallelism for the VJP of VJP of HPE and I'm going to open an issue to fix it (which will be very low-priority). I will also open an issue to document the workaround to build mops-torch without CUDA in environments where a CUDA compiler is available (this PR is a good start, because it adds a MOPS_CUDA option rather than just looking at the availability of a CUDA compiler).

Regarding this PR, I will try to address the review from @Luthaf and then he will be able to review again.

File: mops/include/mops/hpe.hpp [hunk 5]
```diff
@@ -22,7 +31,16 @@ extern template void homogeneous_polynomial_evaluation(
     Tensor<double, 1> output, Tensor<double, 2> A, Tensor<double, 1> C, Tensor<int32_t, 2> indices_A
 );
 
-/// TODO
+/**
+ * @brief Computes the vjp of the homogeneous polynomial evaluation with respect to A.
```
Comment (Luthaf) [inline mops/include/mops/hpe.hpp:35]:
```suggestion
 * @brief Computes the Vector-Jacobian product (vjp) of the homogeneous polynomial evaluation with respect to A.
 *
 * VJP can be used to implement backward propagation of gradients, computing `dL/dA` starting from `dL/dO`, where L is some arbitrary scalar, and O the output of `homogeneous_polynomial_evaluation`.
```

File: mops/include/mops/hpe.hpp [hunk 6]
```diff
@@ -7,7 +7,16 @@
 #include "mops/tensor.hpp"
 
 namespace mops {
-/// TODO
+/**
+ * @brief Computes a homogeneous polynomial on a set of n_batch points.
+ *
```
Comment (Luthaf) [inline mops/include/mops/hpe.hpp:12]: I would also add some math here to show what's being computed

File: mops/include/mops/hpe.hpp [hunk 7]
```diff
@@ -22,7 +31,16 @@ extern template void homogeneous_polynomial_evaluation(
     Tensor<double, 1> output, Tensor<double, 2> A, Tensor<double, 1> C, Tensor<int32_t, 2> indices_A
 );
 
-/// TODO
+/**
+ * @brief Computes the vjp of the homogeneous polynomial evaluation with respect to A.
+ *
+ * @param[out] grad_A The gradient of the output with respect to A, of shape [n_batch, n_factors].
```
Comment (Luthaf) [inline mops/include/mops/hpe.hpp:37]:
```suggestion
 * @param[out] grad_A The gradient of the scalar objective with respect to A, of shape [n_batch, n_factors].
```

File: mops/include/mops/opsaw.hpp [hunk 8]
```diff
@@ -144,6 +195,57 @@ extern template void outer_product_scatter_add_with_weights_vjp(
     Tensor<int32_t, 1> indices_output
 );
 
+/// TODO
```
[the remark this answers is not in the document]
Comment (Luthaf) [reply mops/include/mops/opsaw.hpp:198]: Looks good! I've added a suggestion above to try to make it clearer to people who have no idea what this is about, which you can adapt for all functions!

**Review by Luthaf — APPROVED** [review]

**Comment by Luthaf** [comment]
Let's keep this moving, we can improve the docs further later!

**Pull request merged**